File: src/002/p002.ts
import { fibonacci } from "../utils";

// Simple implementation using a generator
export const solve1 = (): string => {
  let sum = 0;
  let index = 0;
  for (const n of fibonacci()) {
    if (Number(n) > 4000000) {
      break;
    }

    // Add every third term
    index++;
    if (index % 3 === 0) {
      sum += Number(n);
    }
  }
  return sum.toString();
};

export const solve = (): string => {
  let sum = 0;
  for (let a = 1, b = 1; a + b < 4000000; ) {
    const c = a + b;
    a = b + c;
    b = c + a;
    sum += c;
  }
  return sum.toString();
};

File: src/002/p002.py
from typing import Generator
import itertools


def fibonacci(
    a: int = 1, b: int = 1  # pylint: disable=invalid-name
) -> Generator[int, None, None]:
    while True:
        yield a
        a, b = b, a + b


def solve() -> int:
    return sum(
        n for n in itertools.takewhile(lambda n: n < 4000000, fibonacci()) if n % 2 == 0
    )
